uptane-standard.md

```diff
@@ -240,28 +277,222 @@ The Time Server SHALL expose a public interface allowing primaries to communicat
+
+If all three steps complete without error, the ECU SHALL overwrite its current attested time with the time it has just downloaded and generate a new nonce/token for the next request to the time server.
+
+If any check fails, the ECU SHALL NOT overwrite its current attested time, and SHALL jump to the fifth step ({{create_version_report}}). The ECU SHOULD reuse its previous token for the next request to the time server.
+
+#### Verify metadata {#verify_metadata}
+
+The ECU SHALL verify the latest downloaded metadata ({{metadata_verification}}) using either full or partial verification. If the metadata verification fails for any reason, the ECU SHALL jump to the fifth step ({{create_version_report}}).
+
+#### Download latest image {#download_image}
+
+If the ECU does not have secondary storage, it SHALL download the latest image from the primary. (If the ECU has secondary storage, it will already have the latest image in its secondary storage as specified in {{send_images_primary}}, and should skip to the next step.) The ECU MAY first create a backup of its previous working image and store it elsewhere (e.g., the primary).
 
-#### Full verification
+The filename used to identify the latest known image (i.e., the file to request from the primary) SHALL be determined as follows: 
 
-#### Partial verification
+1. Load the targets metadata file from the director repository.
+2. Find the targets metadata associated with this ECU identifier.
+3. Construct the image filename using the rule in {{metadata_filename_rules}}.
 
+When the primary responds to the download request, the ECU SHALL overwrite its current image with the downloaded image from the primary.
 
+If any part of this step fails, the ECU SHALL jump to the fifth step ({{create_version_report}}).
+
+#### Verify image {#verify_image}
+
+The ECU SHALL verify that the latest image matches the latest metadata as follows:
+
+1. Load the latest targets metadata file from the director.
+2. Find the target metadata associated with this ECU identifier.
+3. Check that the hardware identifier in the metadata matches the ECUs hardware identifier.
+4. Check that the release counter of the image in the previous metadata, if it exists, is less than or equal to the release counter in the latest metadata.
+5. If the image is encrypted, decrypt the image with a decryption key to be chosen as follows:
+    * If the ECU key is a symmetric key, the ECU SHALL use the ECU key for image decryption.
+    * If the ECU key is asymmetric, the ECU SHALL check the target metadata for an encrypted symmetric key. If such a key is found, the ECU SHALL decrypt the symmetric key using its ECU key, and use the decrypted symmetric key for image decryption.
+    * If the ECU key is asymmetric and there is no symmetric key in the target metadata, the ECU SHALL use its ECU key for image decryption.
+6. Check that the hash of the image matches the hash in the metadata.
```

Comment:
How does this interact with encrypted firmware? I feel that there might be some complex rules here about whether to validate against the encrypted or unencrypted image:

* If the image is encrypted per ECU, then the Primary can't validate it against the images metadata (because there is one image per ECU).
* If the image is encrypted with an ephemeral key, and that key is sent (encrypted) in the director targets, then the images metadata could contain the hash of the encrypted data. This would mean that the Primary can check it.
* We might want to extend this to deltas in the future. In that case it would be necessary to have separate concepts of 'the size and hash of the final image' (in the images metadata) and 'the size and hash of an encrypted delta' (in the director metadata). It would be possible to check the consistency of the two until the Secondary ECU applied the update.

Reply:
@cajun-rat It doesn't seem like anyone's taken up this question, and I'd like to merge this PR. Can you raise it as an Issue, instead?